captum/optim/_core/output_hook.py

```diff
@@ -130,3 +134,53 @@ def __call__(self, input_t: TupleOfTensorsOrTensorType) -> ModuleOutputMapping:
         finally:
             self.layers.remove_hooks()
         return activations_dict
+
+
+def _remove_all_forward_hooks(
+    module: torch.nn.Module, hook_fn_name: Optional[str] = None
+) -> None:
+    """
+    This function removes all forward hooks in the specified module, without requiring
+    any hook handles. This lets us clean up & remove any hooks that weren't property
+    deleted.
+
+    Warning: Various PyTorch modules and systems make use of hooks, and thus extreme
```

Comment:
As you mentioned this is dangerous to do because we can remove hooks that we didn't set. Why don't we remove only the hooks that we set using hook handles ?

Reply:
The idea was that the problem would be fixed if the individual ran `InputOptimization` a second time, but I now realized that this would break the ability to use multiple instances of `InputOptimization`.

Reply:
Maybe we could somehow let the user decide when to run the cleanup code?

```
[_remove_all_forward_hooks(module, "module_outputs_forward_hook") for module in target_modules]
```

Reply:
There does not appear to be a way to detect whether or not the hooks are still being used, so letting the user decide if they want to perform this fix is probably the best choice.

Reply:
We could probably add this function to the API for users:

```
def cleanup_module_hooks(modules: Union[nn.Module, List[nn.Module]) -> None:
    """
    Remove any InputOptimization hooks from the specified modules. This may be useful
    in the event that something goes wrong in between creating the InputOptimization
    instance and running the optimization function, or if InputOptimization fails
    without properly removing it's hooks.

    Args:

        modules (nn.Module or list of nn.Module): Any module instances that contain
            unused hooks created by InputOptimization, for which the removal of the
            unused hooks is required.
    """
    if not hasattr(modules, "__iter__"):
        modules = [modules]
    # Captum ModuleOutputsHook uses "module_outputs_forward_hook" hook functions
    [_remove_all_forward_hooks(module, "module_outputs_forward_hook") for module in modules]
```

Reply:
@NarineK This main concern with this bug is that it can effect notebook users and other use cases where all the variables are not reset after each attempt.

For example, the following will result in the creation of unused hooks for which the handles required to remove them are no longer accessible:

```
obj = opt.InputOptimization(model, loss_fn, image)
# Failure happens in the InputOptimization init function
```

```
obj = opt.InputOptimization(model, loss_fn, image)
# Failure happens here
history = obj.optimize()
```

```
# Failure happens before the try except block in the optimization function
history = obj.optimize()
```

These unused hooks do not seem to incur any performance penalty as they aren't used, and thus most users can likely ignore them. However, I feel like we should provide the solution for users who's use cases will be impacted by these unused hooks. JIT for example will throw errors when these unused hooks are present.

Reply:
@ProGamerGov, in your example: does it occur after we intialize `opt.InputOptimization(model, loss_fn, image)` second time ? 

Do you have a notebook where we can debug the error ?

Reply:
@NarineK Just setup / install captum with the optim module, and run this snippet of code to reproduce it:


```
!git clone https://github.com/progamergov/captum
%cd captum
!git checkout "optim-wip-fix-hook-bug"
!pip3 install -e .
import sys
sys.path.append('/content/captum')
%cd ..
```

```
import torch
import captum.optim._core.output_hook as output_hook

def test_bug():
    model = torch.nn.Identity()
    for i in range(5):
        _ = output_hook.ModuleOutputsHook([model])
    print(model._forward_hooks.items()) # There will be 5 hooks

test_bug()
```

Reply:
It occurs because we often reuse the same target instance in notebooks for example, and thus the hooks attached to target instance are not removed.

Reply:
cc-ing @vivekmig - Vivek can help with this issue